### DISCUSSION TITLE
Is there any API to get all the registered fields?

### DISCUSSION THREAD

**Discussion by go-mad** [react-hook-form/react-hook-form#3153] [Q&A]
I have a use case where my form fields are initialized based on some dynamic value (let's assume some ID). I want to unregister all the fields when the ID changes. I'm wondering is there any way to either unregister all the fields or to get all the registered fields so that I can unregister them by using a loop.

**Comment by bluebill1049** [comment]
`unregister()` will unregister all the inputs.

**Reply by go-mad** [reply]
Thanks, @bluebill1049  for the quick reply, but the solutions don't work. I still have my old fields registered. Just in case if you have doubt about my code here is the sample snippet:


`useEffect(() => {
    methods.unregister();
  }, [ID]);
`

I'm using nested fields:

Old Fields => `{ parent: {
old_field: "some value"
}}`
New Fields => `{ parent: {
old_field: "some value",
new_field: "some value"
}}`

**Reply by bluebill1049** [reply]
can you reproduce that in a simpe codesandbox? probably input gets re-registered again.

**Reply by go-mad** [reply]
Sure will ping the codesandbox link soon.

**Reply by go-mad** [reply]
@bluebill1049 Here is the codesandbox: https://codesandbox.io/s/nostalgic-sinoussi-vl824?file=/src/App.js

In the example try submitting the first time and check the form data and then try submitting after changing the data. You'll  see in the second submit the old data is also getting attached. The data is logged inside the console.

**Reply by bluebill1049** [reply]
i think you want to use `reset` after the form is submitted
```
useEffect(() => {
  //reset...
}, [isSubmitSucessful])
```

**Reply by go-mad** [reply]
@bluebill1049  No, I don't want to reset. My problem is even after my form data is updated from data1 => data2, after submitting, my final form data is coming as {data1, data2}. I just want to freshly re-register my form when my initial data for the form is updated.

Did you try the example? If you click on the `change data button` and submit it, the form data which is logging is unexpected.

**Reply by bluebill1049** [reply]
yes i did, but if you pass reset(// with your submitted data) isn't that what you after?

**Reply by go-mad** [reply]
No, the problem is submitted data is not even correct. I shouldn't be getting the `both` object in case of `data2`. In the screenshot I didn't submit `data1` I just changed the data and submitted `data2` 
![Screenshot 2020-10-12 at 3 43 06 PM](https://user-images.githubusercontent.com/53913590/95734852-a2786800-0ca1-11eb-855d-4585234ad264.png)

**Reply by bluebill1049** [reply]
hmm looks like unregister() doesn't support clean all inputs... https://codesandbox.io/s/mystifying-bash-zr8od?file=/src/App.tsx

**Reply by bluebill1049** [reply]
happy for you to submit a feature request on this, i will get that sorted for you with `unregister()` reset entire form.

**Reply by bluebill1049** [reply]
i will add to this board to feature, see if others need this feature, and prioritise when to deliver it.

**Reply by go-mad** [reply]
@bluebill1049 Sure please add it.
